### ISSUE TITLE
Servo WebRenderer for Xi front-end

### ISSUE THREAD

**Issue by Kerollmops** [google/xi-editor#140]
I like the way you think, first goal: performances.
But I think about something else: performances, multi-OS and customization.
That's just something I think about and I believe I'm not the only one.

Why not opening a thread to speak about this ? 😄

**Comment by chriskrycho** [comment]
A pure-web front-end built on something like Servo is a very neat idea, and I think definitely something to explore. Especially with something like Elm, it could be *very* performant; I'm certainly interested in the possibilities long-term for something like that, which uses the DOM context *only* for the rendering and does all the hard work in Rust.

**Comment by myfreeweb** [comment]
You can use WebRender directly, without Servo — no need for the bloated web stuff (DOM, HTML, CSS, JS). There's [a sample](https://github.com/servo/webrender/blob/master/sample/src/main.rs) in the webrender repo.

**Comment by chriskrycho** [comment]
Some of that bloated web stuff would actually be useful for building an editor front-end, if you did not want to re-implement all of those primitives yourself. 😜

**Comment by myfreeweb** [comment]
Of course it can be *useful*, but it's not necessary, performant or lightweight. The Xi backend/core is extremely focused on performance — the frontends should as well.

Now that I think about it, WebRender isn't necessary either. It's good for rendering complex layouts — a text editor is mostly, uh, *text*. [The glium backend](https://github.com/potocpav/xi_glium) is the right idea IMO.

**Comment by chriskrycho** [comment]
While I do not disagree a bit about the aim of high-performance front ends, there is also a great deal to be said for the accessibility of web technologies. Coupling that to a high-performance back end might yield some interesting results. And if someone wants to do that, it doesn't hurt projects like the gloom back end a bit.

**Comment by thomas-profitt** [comment]
I see absolutely no reason _not_ to create more front-ends with whatever technologies people are interested in using to create front-ends.

That feeds back into xi-core development. It helps us get the communication between the front-end and core to shape up, and it helps the plugin architecture take shape as well.

Right now, while all that stuff is still changing and developing, the value of front-ends is almost exclusively limited to helping development of xi-core and how front-ends, core and plugins communicate. With that in mind, _more front-ends_!

**Comment by ConnorGBrewster** [comment]
I am interested in getting involved with this project and already had the thought of building a ui library of some sort on top of webrender to allow for easy cross platform desktop UIs (I am a contributor to the Servo project). Would something like this be of interest?

**Comment by disordinary** [comment]
Don't get me wrong, I think this is a great idea and will be following the development with great interest. But isn't the stated point of the project to have native UI? Otherwise it's just re-creating Atom.

If, despite this, a cross platform solution is considered to be a good idea then using the webstack is a great option because it cuts the development time, gives advanced layout and animation support out of the box (which will improve as Servo improves), and gives a highly customisable, flexible, hackable, and themable interface. 

But then why go through all the effort of using Servo? The way that Xi is written it can interop with any language, so an electron or NW.js based editor could communicate with Xi the same way a Swift based editor can, and if you want to build most of the code in Rust then compiling to asm.js via emscripten or writing a plugin for the node runtime embedded in electron or nwjs via https://github.com/neon-bindings/neon are both options.

Despite playing devils advocate above, I'd love to see a servo based application framework similar to NWjs or Electron emerge from the Swift team / rust community.

**Comment by ConnorGBrewster** [comment]
@disordinary I agree. I was not suggesting to build a frontend built on servo or a web stack as that has already been done and would seem to defeat the purpose of this project. 

I believe that native ui goal is an awesome goal and should continue to be one; however, it would also be nice to have an alternate frontend that looks and behaves exactly the same on every platform while maintaining the performance gains of a Rust core and minimal stack. Using a tool like webrender would allow for advanced styling and high performance as it can be used completely independent of servo and any web stack. It just renders display lists that are generated from the layout/style engine in servo.

**Comment by disordinary** [comment]
But a cross platform run time that has a closer coupling between the interpreted code and the layout engine / renderer could be really interesting.
